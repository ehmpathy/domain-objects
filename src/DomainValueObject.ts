import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Value Object is a type of Domain Object for which:
 * - is immutable (i.e, properties do not change over time)
 * - is defined by its properties (i.e., it is uniquely identified by the composition of all of its properties)
 *
 * For example,
 * - A `Address { street, city, state, country }` is a value object. Changing any of the properties, like `street`, produces a completely new address
 * - A `Name { first, last }` is a value object. Changing either property means you are dealing with a new name.
 */
export abstract class DomainValueObject<T> extends DomainObject<T> {}
