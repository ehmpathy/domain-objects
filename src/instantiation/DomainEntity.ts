import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, an Entity is a type of Domain Object for which:
 * - properties change over time (e.g., it has a life cycle)
 * - identity matters (e.g., two entities could have the same properties, differing only by id, and still be valid)
 *
 * For example,
 * - A `User { uuid, name, address, profilePicture }` is an entity. Properties like `name`, `address`, and `profilePicture` can change over time.
 * - A `Job { uuid, requestId, status }` is an entity. We expect `status` to change over time as the job goes from, for example, `QUEUED` to `FULFILLED`.
 */
export class DomainEntity<T> extends DomainObject<T> {
  /**
   * `DomainEntity.unique` defines all of the properties of the entity that the entity is naturally unique on
   *
   * For example,
   * - a `FileProcessingJob { uuid, filePath, status }` is likely going to be unique on the `filePath` if we only ever need to process the same file once.
   * - a `SupportTicket { uuid, userId, type, status }`, on the other hand, is likely only going to be unique on the `uuid`, because the same user can create the same support ticket over and over again
   */
  public static unique: readonly string[];

  /**
   * `DomainEntity.updatable` defines all of the properties of the entity that are updatable
   *
   * For example,
   * - a `User { uuid, name, address, picture }` would likely have `unique = ['name', 'address', 'picture']`, to enable the user to update their information as it changes over time
   * - an `Email { uuid, externalId, toAddress, fromAddress, message, status }` is likely only going to have `status` be updatable, as by changing the `toAddress` would, in this case, make us consider it to be a different email!
   */
  public static updatable: readonly string[];
}
