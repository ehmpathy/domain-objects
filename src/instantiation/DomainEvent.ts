import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, an Event is a type of Domain Object for which:
 * - properties are immutable (e.g., something happened)
 * - identity does matter
 *   - i.e., it is uniquely identified by what happened and to what
 *   - i.e., if you change the uniquely identifying properties (when or to what) of an event, it will be considered a new event
 *   - i.e., not all properties on the event uniquely identify the event
 *
 * The purpose of a Domain Event is to represent interesting things that happen in the domain.
 *
 * For example,
 * - An `ItemViewedEvent { itemUuid, customerUuid, occurredAt }` is an event. Changing when it `occurredAt` which `customerUuid` did it or which `itemUuid` it happened to would result in a new event.
 * - A `PowerSupplyMeasuredEvent { batteryUuid, current, voltage, temperature, occurredAt }` is an event. Changing when it `occurredAt` or which `batteryUuid` it occurred to would result in a new event. Note: changing the `current`, `voltage`, or `temperature` would _not_ result in a new event though, since only one measurement can be made at a time for a `batteryUuid`
 */
export abstract class DomainEvent<
  T extends Record<string, any>,
> extends DomainObject<T> {
  /**
   * `DomainEvent.unique` defines all of the properties of the event that the event is naturally unique on.
   *
   * It should always include a timestamp.
   *
   * For example,
   * - a `FileOpenedEvent { fileUuid, userUuid, occurredAt }` is likely going to be unique on the all of the keys: the combination of `['fileUuid', 'userUuid', 'occurredAt']` uniquely identifies the event.
   * - an `AirComfortMeasuredEvent { locationUuid, sensorUuid, temperature, humidity, occurredAt }`, on the other hand, is likely only going to be unique on `['locationUuid', 'sensorUuid', 'occurredAt']`, because logically we cant have a measurement from the same `sensor` + `location` at the same `time`
   */
  public static unique: readonly string[];
}
