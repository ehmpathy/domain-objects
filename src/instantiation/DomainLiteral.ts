import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Literal (a.k.a. Value Object), is a type of Domain Object for which:
 * - properties are immutable
 *   - i.e., it represents some literal value which happens to have a structured object shape
 *   - i.e., if you change the value of any of its properties, it is a different literal
 * - identity does not matter
 *   - i.e., it is uniquely identifiable by its non-metadata properties
 *
 * The purpose of a Domain Literal is to represent distinct, literal values which happen to have a structured object shape
 *
 * For example,
 * - A `Address { street, city, state, country }` is a literal. Changing any of the properties, like `street`, produces a completely new address
 * - A `Geocode { latitude, longitude }` is a literal. Changing either property means you are dealing with a new geocode
 */
export abstract class DomainLiteral<
  T extends Record<string, any>,
> extends DomainObject<T> {}
