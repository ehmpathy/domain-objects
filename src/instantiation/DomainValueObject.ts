import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Value Object is a type of Domain Object for which:
 * - properties are immutable (e.g., its a fact)
 * - identity does not matter
 *   - i.e., it is uniquely identifiable by its properties
 *   - i.e., if you change the value of any of its properties, it is now considered a new value-object
 *
 * For example,
 * - A `Address { street, city, state, country }` is a value object. Changing any of the properties, like `street`, produces a completely new address
 * - A `Name { first, last }` is a value object. Changing either property means you are dealing with a new name.
 */
export abstract class DomainValueObject<T> extends DomainObject<T> {}
