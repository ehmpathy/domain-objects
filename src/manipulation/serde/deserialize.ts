/* eslint-disable no-underscore-dangle */
/* eslint-disable @typescript-eslint/no-use-before-define */
import { DomainObject } from '../..';

export class DeserializationMissingDomainObjectClassError extends Error {
  constructor({ className }: { className: string }) {
    const message = `
DomainObject '${className}' was referenced in the string being deserialized but was missing from the context given to the deserialize method.

Please make sure all DomainObjects serialized in the string have their classes defined in the context given to the deserialize method, using the 'with' property.
    `;
    super(message);
  }
}

/**
 * Revives the domain objects stored in a string produced by the `serialize` method
 *
 * Use Cases:
 * - persistance
 *   - e.g., deserialize domain objects from a string that was saved to a persistant store (i.e., a string produced by the `serialize` method)
 */
export const deserialize = <T extends any>(serialized: string, context: { with?: DomainObject<any>[] } = {}): T => {
  // parse the string
  const parsed = JSON.parse(serialized);

  // recursively traverse the parsed value to hydrate all domain objects
  return toHydrated(parsed, { with: context.with ?? [] });
};

/**
 * helper method for deserialize
 *
 * hydrates any domain-objects in the previously serialized value, if present
 */
const toHydrated = (value: any, context: { with: DomainObject<any>[] }): any => {
  // if this value is not an array and is not an object, then it's a literal, and there's no more hydration to be done
  if (!Array.isArray(value) && typeof value !== 'object') return value;

  // if its null, return it too (null is typeof 'object' in js :shrug:)
  if (value === null) return null;

  // if its an array, then `toHydrated` each element
  if (Array.isArray(value)) return value.map((el) => toHydrated(el, context));

  // otherwise, hydrate the object
  return toHydratedObject(value, context);
};

/**
 * helper method for `toSerializable`
 *
 * converts all objects to deterministically serializable objects (e.g., sort keys, handle domain objects, etc)
 */
const toHydratedObject = (obj: Record<string, any>, context: { with: DomainObject<any>[] }) => {
  // if this object is a domain object, lookup its constructor and hydrate it
  if (obj._dobj) {
    // pull its name from the key that the `serialize` method sticks the name onto
    const domainObjectClassName = obj._dobj;

    // lookup the domain object constructor from context
    const DomainObjectConstructor = context.with.find((constructor) => (constructor as typeof DomainObject).name === domainObjectClassName) as
      | typeof DomainObject
      | undefined;
    if (!DomainObjectConstructor) throw new DeserializationMissingDomainObjectClassError({ className: domainObjectClassName });

    // hydrate the domain object, now that it was given.
    return new DomainObjectConstructor(obj); // (note: domain objects hydrate their nested domain-object properties themselves, so we can just return the result here :smile:)
  }

  // since this was not a domain object, recursively traverse each key
  const hydratedObj: Record<string, any> = {};

  // for each key, recursively hydrate it
  Object.keys(obj).forEach((key) => {
    hydratedObj[key] = toHydrated(obj[key], context);
  });

  // return it
  return hydratedObj;
};
