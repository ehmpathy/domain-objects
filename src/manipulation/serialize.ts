/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable no-underscore-dangle */
import { assertDomainObjectIsSafeToManipulate } from '../constraints/assertDomainObjectIsSafeToManipulate';
import { DomainEntity } from '../instantiation/DomainEntity';
import { DomainObject } from '../instantiation/DomainObject';
import { DomainValueObject } from '../instantiation/DomainValueObject';
import { getUniqueIdentifier } from './getUniqueIdentifier';

/**
 * Returns a deterministic, human readable, and de-serializable string representation of a Domain Object.
 *
 * Deterministic:
 * - orders object keys
 * - sorts elements in arrays
 * - replaces nested domain objects with their minimal and static representation (i.e., calls `getUniqueIdentifier` on any key's value w/ instance of DomainEntity | DomainValueObject)
 *
 * Use Cases:
 * - change detection
 *   - e.g., determine whether there has been a change to an entity or not, by serializing the previous and current versions, and comparing equality
 *   - i.e., `const areEqual = serialize(userBefore) === serialize(userNow);`
 *
 * - identity comparison
 *   - e.g., determine whether two domain objects are the same, by getting their unique identities, serializing each and comparing equality
 *   - i.e., `const areEqual = serialize(getUniqueIdentity(contactMethod1)) === serialize(getUniqueIdentity(contactMethod2))`
 */
export const serialize = (value: any): string => {
  // 1. convert the value to a deterministically serializable object
  const serializableValue = toSerializable(value, true);

  // 2. json stringify it last, so that we dont have a recursively growing set of nested quotes
  return JSON.stringify(serializableValue);
};

/**
 * helper method for serialize
 *
 * converts any value to a deterministically serializable representation (i.e., sorts arrays + object keys, casts .toString(), handles domain objects, etc)
 */
export const toSerializable = (value: any, root?: boolean): any => {
  // if this value is not an array and has a `.toString` definition, return it itself
  if (!Array.isArray(value) && typeof value !== 'object') return value;

  // if its null, return it too (null is typeof 'object' in js :shrug:)
  if (value === null) return null;

  // if its an array, then `toSerializable` each element, and sort the result
  if (Array.isArray(value)) return value.map((el) => toSerializable(el)).sort((a, b) => (JSON.stringify(a) < JSON.stringify(b) ? -1 : 1)); // stringify to sort, to sort objects correctly

  // if it has a `.toString()` method, use the `.toString()` method; (e.g., new Date().toString())
  if (value.toString()) {
    const toStringResult = value.toString();
    if (toStringResult !== '[object Object]') return toStringResult; // use the result, unless its the "generic object -> ['object Object']" result, which isn't helpful
  }

  // otherwise, prepare to make the object stringifiable
  return toSerializableObject(value, root);
};

/**
 * helper method for `toSerializable`
 *
 * converts all objects to deterministically serializable objects (e.g., sort keys, handle domain objects, etc)
 */
const toSerializableObject = (obj: Record<string, any>, root?: boolean) => {
  const stringifiableObj: Record<string, any> = {};

  // if this object is a DomainObject, add its name to the serial parts as metadata (i.e., underscore suffix)
  if (obj instanceof DomainObject) stringifiableObj._dobj = obj.constructor.name;

  // if this object is a domain object, make sure that it is safe to manipulate
  if (obj instanceof DomainObject) assertDomainObjectIsSafeToManipulate(obj);

  // if this object is a persistable DomainObject AND its not the root object we're serializing, then only consider its unique properties for serialization; (i.e., only consider which domain objects the root object references, not the current state of the domain objects the root object references)
  let objToMakeSerializable = obj;
  if (!root && (obj instanceof DomainEntity || obj instanceof DomainValueObject)) {
    objToMakeSerializable = getUniqueIdentifier(obj);
  }

  // for each key, in sorted order, get the serial representation
  Object.keys(objToMakeSerializable)
    .sort()
    .forEach((key) => {
      stringifiableObj[key] = toSerializable(obj[key]);
    });

  // return it
  return stringifiableObj;
};
