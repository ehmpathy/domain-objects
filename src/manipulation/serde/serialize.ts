/* eslint-disable @typescript-eslint/no-use-before-define */

/* eslint-disable no-underscore-dangle */
import { assertDomainObjectIsSafeToManipulate } from '../../constraints/assertDomainObjectIsSafeToManipulate';
import { DomainEntity } from '../../instantiation/DomainEntity';
import { DomainLiteral } from '../../instantiation/DomainLiteral';
import { DomainObject } from '../../instantiation/DomainObject';
import { getUniqueIdentifier } from '../getUniqueIdentifier';

interface SerializeOptions {
  /**
   * set this to true if you need to deserialize the domain objects losslessly
   *
   * note
   * - this will make it so we do not call `getUniqueIdentifier` on the nested domain objects
   * - by default, this is set to false, so that this can be used for change detection
   */
  lossless?: boolean;

  /**
   * set this to true if the order of items in arrays does not matter in your usecase
   *
   * note
   * - this will make it so that the original order of items in arrays is not preserved
   * - by default, this is set to false, since order matters in arrays by common definition
   */
  orderless?: boolean;
}

/**
 * Returns a deterministic, human readable, and de-serializable string representation of a Domain Object.
 *
 * Deterministic:
 * - orders object keys
 * - sorts elements in arrays
 * - when lossless=false, replaces nested domain objects with their minimal and static representation (i.e., calls `getUniqueIdentifier` on any nested domain objects)
 *
 * Use Cases:
 * - change detection
 *   - e.g., determine whether there has been a change to an entity or not, by serializing the previous and current versions, and comparing equality
 *   - i.e., `const areEqual = serialize(omitMetadataValues(userBefore)) === serialize(omitMetadataValues(userNow));`
 *
 * - identity comparison
 *   - e.g., determine whether two domain objects are the same, by getting their unique identities, serializing each and comparing equality
 *   - i.e., `const areEqual = serialize(getUniqueIdentity(contactMethod1)) === serialize(getUniqueIdentity(contactMethod2))`
 *
 * - persistance
 *   - e.g., serialize domain objects into a persistant store in string format (to be later revived with the `deserialize` method)
 */
export const serialize = (
  value: any,
  options: SerializeOptions = { lossless: false, orderless: false },
): string => {
  // convert the value to a deterministically serializable object
  const serializableValue = toSerializable(value, options, true);

  // json stringify it last, so that we dont have a recursively growing set of nested quotes
  return JSON.stringify(serializableValue);
};

/**
 * helper method for serialize
 *
 * converts any value to a deterministically serializable representation (i.e., sorts arrays + object keys, casts .toString(), handles domain objects, etc)
 */
const toSerializable = (
  value: any,
  options: SerializeOptions,
  root?: boolean,
): any => {
  // if this value is not an array and is not an object, then it's a literal, so no more preparation required, return it itself
  if (!Array.isArray(value) && typeof value !== 'object') return value;

  // if its null, return it too (null is typeof 'object' in js :shrug:)
  if (value === null) return null;

  // if its an array, then `toSerializable` each element, and sort the result if needed
  if (Array.isArray(value))
    return value
      .map((el) => toSerializable(el, options))
      .sort((a, b) => {
        // if order was said to not matter, then sort in ascending value of their stringified form, to allow for deterministic comparisons
        if (options.orderless)
          return JSON.stringify(a) < JSON.stringify(b) ? -1 : 1;

        // otherwise, retain original order
        return 0;
      });

  // if it has a `.toString()` method, use the `.toString()` method; (e.g., new Date().toString())
  if (value.toString()) {
    const toStringResult = value.toString();
    if (toStringResult !== '[object Object]') return toStringResult; // use the result, unless its the "generic object -> ['object Object']" result, which isn't helpful
  }

  // otherwise, prepare to make the object stringifiable
  return toSerializableObject(value, options, root);
};

/**
 * helper method for `toSerializable`
 *
 * converts all objects to deterministically serializable objects (e.g., sort keys, handle domain objects, etc)
 */
const toSerializableObject = (
  obj: Record<string, any>,
  options: SerializeOptions,
  root?: boolean,
) => {
  const stringifiableObj: Record<string, any> = {};

  // if this object is a DomainObject, add its name to the serial parts as metadata (i.e., underscore suffix)
  if (obj instanceof DomainObject)
    stringifiableObj._dobj = obj.constructor.name;

  // if this object is a domain object, make sure that it is safe to manipulate
  if (obj instanceof DomainObject) assertDomainObjectIsSafeToManipulate(obj);

  // if this object is a persistable DomainObject AND its not the root object we're serializing && lossless is off, then only consider its unique properties for serialization; (i.e., only consider which domain objects the root object references, not the current state of the domain objects the root object references)
  let objToMakeSerializable = obj;
  if (
    !root &&
    !options.lossless &&
    (obj instanceof DomainEntity || obj instanceof DomainLiteral)
  ) {
    objToMakeSerializable = getUniqueIdentifier(obj);
  }

  // for each key, in sorted order, get the serial representation
  Object.keys(objToMakeSerializable)
    .sort()
    .forEach((key) => {
      stringifiableObj[key] = toSerializable(obj[key], options);
    });

  // return it
  return stringifiableObj;
};
