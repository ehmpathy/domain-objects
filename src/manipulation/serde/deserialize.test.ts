import { DomainEntity } from '../../instantiation/DomainEntity';
import { DomainValueObject } from '../../instantiation/DomainValueObject';
/* eslint-disable no-useless-escape */
import { serialize } from './serialize';
import { deserialize } from './deserialize';

describe('deserialize', () => {
  describe('basic types', () => {
    it('should deserialize strings', () => {
      const original = 'hello!';
      const serial = serialize(original);
      const undone = deserialize(serial);
      expect(undone).toEqual(original);
    });
    it('should deserialize numbers', () => {
      const original = 821;
      const serial = serialize(original);
      const undone = deserialize(serial);
      expect(undone).toEqual(original);
    });
    it.todo('should deserialize dates');
    it.todo('should deserialize undefined');
    it('should deserialize nulls', () => {
      const original = null;
      const serial = serialize(original);
      const undone = deserialize(serial);
      expect(undone).toEqual(original);
    });
    it.todo('should deserialize buffers');
  });
  describe('arrays', () => {
    it('should be able to deserialize arrays', () => {
      const original = ['2', 'four', 1, 3];
      const serial = serialize(original);
      const undone = deserialize(serial);
      expect(undone).toEqual(original); // sorted
    });
    it('should deserialize arrays even if they have objects', () => {
      const original = ['banana', 821, { id: 0, meaning: null }, { id: 1, meaning: 42, value: 821 }];
      const serial = serialize(original);
      const undone = deserialize(serial);
      expect(undone).toEqual(original);
    });
  });
  describe('objects', () => {
    it('should be able to serialize an object with all sorts of types', () => {
      const original = {
        color: 'blue',
        cost: 821,
        orders: [
          { id: 0, meaning: null },
          { id: 1, meaning: 42, value: 821 },
        ],
        application: {
          type: 'PAINTING',
        },
      };
      const serial = serialize(original);
      const undone = deserialize(serial);
      expect(undone).toEqual(original);
    });
  });
  describe('domain objects', () => {
    // define some domain objects to test with
    interface Spaceship {
      serialNumber: string;
      fuelQuantity: number;
      passengers: number;
    }
    class Spaceship extends DomainEntity<Spaceship> implements Spaceship {
      public static unique = ['serialNumber'];
      public static updatable = ['serialNumber'];
    }
    interface Address {
      id?: number;
      galaxy: string;
      solarSystem: string;
      planet: string;
      continent: string;
    }
    class Address extends DomainValueObject<Address> implements Address {}
    interface Spaceport {
      uuid: string;
      address: Address;
      spaceships: Spaceship[];
    }
    class Spaceport extends DomainEntity<Spaceport> implements Spaceport {
      public static unique = ['uuid'];
      public static updatable = ['spaceships'];
      public static nested = { address: Address, spaceships: Spaceship };
    }

    // run the tests
    it('should deserialize domain objects', () => {
      const ship = new Spaceship({
        serialNumber: '__UUID__',
        fuelQuantity: 9001,
        passengers: 21,
      });
      const original = ship;
      const serial = serialize(original);
      const undone = deserialize(serial, { with: [Spaceship] });
      expect(undone).toEqual(original);
      expect(undone).toBeInstanceOf(Spaceship);
    });
    it('should throw a helpful error if attempted to deserialize a domain object without its constructor being provided in the context', () => {
      const ship = new Spaceship({
        serialNumber: '__UUID__',
        fuelQuantity: 9001,
        passengers: 21,
      });
      const original = ship;
      const serial = serialize(original);
      try {
        deserialize(serial, { with: [] });
        throw new Error('should not reach here');
      } catch (error) {
        expect(error.message).toContain(
          `DomainObject 'Spaceship' was referenced in the string being deserialized but was missing from the context given to the deserialize method`,
        );
        expect(error.message).toMatchSnapshot(); // save an example of the message to snapshot
      }
    });
    it('recursively deserialize domain objects', () => {
      const shipA = new Spaceship({
        serialNumber: '__SHIP_A__',
        fuelQuantity: 7000,
        passengers: 42,
      });
      const shipB = new Spaceship({
        serialNumber: '__SHIP_B__',
        fuelQuantity: 9001,
        passengers: 21,
      });
      const spaceport = new Spaceport({
        uuid: '__SPACEPORT_UUID__',
        address: new Address({ galaxy: 'Milky Way', solarSystem: 'Sun', planet: 'Earth', continent: 'North America' }),
        spaceships: [shipA, shipB],
      });
      const original = spaceport;
      const serial = serialize(original, { lossless: true });
      const undone = deserialize<Spaceport>(serial, { with: [Spaceport, Spaceship, Address] });
      expect(undone).toEqual(original);
      expect(undone).toBeInstanceOf(Spaceport);
      expect(undone.address).toBeInstanceOf(Address);
      expect(undone.spaceships[0]).toBeInstanceOf(Spaceship);
    });
  });
});
