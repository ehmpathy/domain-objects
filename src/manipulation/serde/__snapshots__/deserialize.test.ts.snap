// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deserialize domain objects should throw a helpful error if attempted to deserialize a domain object without its constructor being provided in the context 1`] = `
"
DomainObject 'Spaceship' was referenced in the string being deserialized but was missing from the context given to the deserialize method.

Please make sure all DomainObjects serialized in the string have their classes defined in the context given to the deserialize method, using the 'with' property.
    "
`;
